pkg/operator/controller/externaldns/credentials-request_test.go

```diff
@@ -0,0 +1,287 @@
+		errExpected               bool
+		ocpPlatform               bool
+		inputExtDNS               *operatorv1alpha1.ExternalDNS
+		expectedCredentialRequest *cco.CredentialsRequest
+	}{
+		{
+			name:                      "Ensure Credential request when no credential requests are there in AWS",
+			existingObjects:           []runtime.Object{testAWSExtDNSInstanceWhenOCPRouteSourceWhenAWSCredentialsNotProvided(), testSrcSecretWhenSourceOCP()},
+			inputExtDNS:               testAWSExtDNSInstanceWhenOCPRouteSourceWhenAWSCredentialsNotProvided(),
+			expectedCredentialRequest: getCredentialRequest(controller.ExternalDNSCredentialsRequestName(testAWSExtDNSInstanceWhenOCPRouteSourceWhenAWSCredentialsNotProvided()), testAWSExtDNSInstanceWhenOCPRouteSourceWhenAWSCredentialsNotProvided().Spec.Provider.Type),
+		},
+		{
+			name:                      "Ensure Credential request when no credential requests are there in Azure",
+			existingObjects:           []runtime.Object{testAzureExtDNSInstanceWhenSourceIsOCPRoute(), testSrcSecretWhenSourceOCP()},
+			inputExtDNS:               testAzureExtDNSInstanceWhenSourceIsOCPRoute(),
+			expectedCredentialRequest: getCredentialRequest(controller.ExternalDNSCredentialsRequestName(testAzureExtDNSInstanceWhenSourceIsOCPRoute()), testAzureExtDNSInstanceWhenSourceIsOCPRoute().Spec.Provider.Type),
+		},
+		{
+			name:                      "Ensure Credential request when no credential requests are there in GCP",
+			existingObjects:           []runtime.Object{testGCPExtDNSInstanceWhenSourceIsOCPRoute(), testSrcSecretWhenSourceOCP()},
+			inputExtDNS:               testGCPExtDNSInstanceWhenSourceIsOCPRoute(),
+			expectedCredentialRequest: getCredentialRequest(controller.ExternalDNSCredentialsRequestName(testGCPExtDNSInstanceWhenSourceIsOCPRoute()), testGCPExtDNSInstanceWhenSourceIsOCPRoute().Spec.Provider.Type),
+		},
+	}
+	for _, tc := range testCases {
+		cl := fake.NewClientBuilder().WithScheme(testScheme).WithRuntimeObjects(tc.existingObjects...).Build()
+
+		r := &reconciler{
+			client: cl,
+			scheme: testScheme,
+			config: testConfig(),
+			log:    zap.New(zap.UseDevMode(true)),
+		}
+
+		t.Run(tc.name, func(t *testing.T) {
+
+			_, got, _ := r.ensureExternalCredentialsRequest(context.TODO(), tc.inputExtDNS)
+
+			if reflect.DeepEqual(got.Name, tc.expectedCredentialRequest.Name) == false {
+				t.Errorf("Got name %v  but expected name is %v", got.Name, tc.expectedCredentialRequest.Name)
+			}
+
+			if reflect.DeepEqual(got.Namespace, tc.expectedCredentialRequest.Namespace) == false {
+				t.Errorf("Got namespace %v  but expected namespace is %v", got.Namespace, tc.expectedCredentialRequest.Namespace)
+			}
+
+			if tc.inputExtDNS.Spec.Provider.Type == operatorv1alpha1.ProviderTypeAWS {
+				codec, _ := cco.NewCodec()
+				awsSpec := cco.AWSProviderSpec{}
+				err := codec.DecodeProviderSpec(got.Spec.ProviderSpec, &awsSpec)
+				if err != nil {
+					t.Errorf("error decoding provider v1 spec: %v", err)
+				}
```

Comment:
Can you please factorize the decoding logic into a dedicated function? This would save a lot of space.

Reply:
factorizing wont help here as  codec.DecodeProviderSpec expects provider specific Spec. it is very hard to explain here what the problem is. If you have some ideas then please explain. But this a very simple thing  like having a function with switch for provider types. Do you want a seperate functions for different providers ?